File: 백준/Gold/14502. 연구소/연구소.py
n, m = map(int, input().split())
a = []
for _ in range(n):
    a.append(list(input().split()))

virus = []
wall = []
for i in range(n):
    for j in range(m):
        if a[i][j] == '2':
            virus.append((i, j))
        elif a[i][j] == '0':
            wall.append((i, j))

dx = [-1, 0, 1, 0]
dy = [0, 1, 0, -1]
visited = list()


def dfs(x, y):
    for i in range(4):
        nx, ny = x + dx[i], y + dy[i]
        if nx not in range(n) or ny not in range(m): continue
        if visited[nx][ny]: continue
        if a[nx][ny] == '1': continue
        visited[nx][ny] = True
        dfs(nx, ny)


def go():
    global visited

    count = 0
    visited = [[False] * m for _ in range(n)]

    for v in virus:
        visited[v[0]][v[1]] = True
        dfs(v[0], v[1])

    for i in range(n):
        for j in range(m):
            if a[i][j] == '0' and not visited[i][j]:
                count += 1

    return count


# 벽 3개 세우기
answer = 0
for i in range(len(wall)):
    for j in range(i):
        for k in range(j):
            a[wall[i][0]][wall[i][1]] = '1'
            a[wall[j][0]][wall[j][1]] = '1'
            a[wall[k][0]][wall[k][1]] = '1'
            answer = max(answer, go())
            a[wall[i][0]][wall[i][1]] = '0'
            a[wall[j][0]][wall[j][1]] = '0'
            a[wall[k][0]][wall[k][1]] = '0'

print(answer)

File: 백준/Gold/14502. 연구소/연구소.kt
fun main(args: Array<String>) {
    // 0: 빈칸, 1: 벽, 2: 바이러스
    // 무조건 3개의 벽을 세워야 함
    val (n, m) = readln().split(" ").map { it.toInt() }
    val map = Array(n) { IntArray(m) }
    val viruses = mutableListOf<Pair<Int, Int>>()
    val walls = mutableListOf<Pair<Int, Int>>()
    var answer = 0
    repeat(n) { r ->
        map[r] = readln().split(" ").map { it.toInt() }.toIntArray().also { row ->
            row.indices.forEach { i ->
                when (row[i]) {
                    0 -> walls.add(r to i)
                    2 -> viruses.add(r to i)
                }
            }
        }
    }

    val dx = listOf(-1, 0, 1, 0)
    val dy = listOf(0, 1, 0, -1)
    var visited = Array(n) { BooleanArray(m) }

    fun dfs(x: Int, y: Int) {
        for (i in 0 until 4) {
            val nx = x + dx[i]
            val ny = y + dy[i]
            if (nx !in 0 until m || ny !in 0 until n) continue
            if (visited[ny][nx] || map[ny][nx] == 1) continue
            visited[ny][nx] = true
            dfs(nx, ny)
        }
    }

    fun solve(): Int {
        var count = 0
        visited = Array(n) { BooleanArray(m) }

        viruses.forEach { v ->
            visited[v.first][v.second] = true
            dfs(v.second, v.first)
        }

        for (i in 0 until n) {
            for (j in 0 until m) {
                if (map[i][j] == 0 && !visited[i][j]) count++
            }
        }
        return count
    }

    // 1. 벽을 세운다
    for (i in walls.indices) {
        for (j in 0 until i) {
            for (k in 0 until j) {
                map[walls[i].first][walls[i].second] = 1
                map[walls[j].first][walls[j].second] = 1
                map[walls[k].first][walls[k].second] = 1
                answer = maxOf(answer, solve())
                map[walls[i].first][walls[i].second] = 0
                map[walls[j].first][walls[j].second] = 0
                map[walls[k].first][walls[k].second] = 0
            }
        }
    }
    println(answer)
}